File: backend/app/api/quotes.py
from uuid import UUID
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from pydantic import BaseModel

from app.core.deps import (
    get_current_admin,
    get_current_user,
    get_db,
    get_pagination,
)
from app.models.user import User
from app.schemas.quote import QuoteCreate, QuoteListResponse, QuotePublic, QuoteUpdate
from app.services.quote_service import (
    create_quote,
    get_quote,
    list_quotes,
    set_status,
    update_quote,
)


class StatusChange(BaseModel):
    status: Literal["SENT", "APPROVED", "REJECTED", "EXPIRED"]


router = APIRouter(prefix="/sales/quotes", tags=["sales-quotes"])


@router.get("", response_model=QuoteListResponse)
def list_quotes_endpoint(
    pagination: tuple[int, int] = Depends(get_pagination),
    search: str | None = None,
    status: str | None = None,
    partner_id: UUID | None = None,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    page, page_size = pagination
    items, total = list_quotes(db, page, page_size, search, status, partner_id)
    return {"items": items, "total": total, "page": page, "page_size": page_size}


@router.get("/{quote_id}", response_model=QuotePublic)
def get_quote_endpoint(
    quote_id: UUID,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    quote = get_quote(db, quote_id)
    if not quote:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Quote not found")
    return quote


@router.post("", response_model=QuotePublic, status_code=status.HTTP_201_CREATED)
def create_quote_endpoint(
    data: QuoteCreate,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin),
):
    quote = create_quote(db, data)
    return quote


@router.put("/{quote_id}", response_model=QuotePublic)
def update_quote_endpoint(
    quote_id: UUID,
    data: QuoteUpdate,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin),
):
    try:
        quote = update_quote(db, quote_id, data)
    except ValueError:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="immutable_state")
    if not quote:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Quote not found")
    return quote


@router.post("/{quote_id}/status", response_model=QuotePublic)
def change_status_endpoint(
    quote_id: UUID,
    body: StatusChange,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin),
):
    try:
        quote = set_status(db, quote_id, body.status)
    except ValueError:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="invalid_status")
    if not quote:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Quote not found")
    return quote

File: web/src/lib/quotes.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import api from './api'
import { Quote, QuoteDetail, QuoteCreate, QuoteUpdate } from '../types/quote'

interface ListParams {
  search: string
  status: string
  partnerId: string
  page: number
  pageSize: number
}

export const listQuotes = ({ search, status, partnerId, page, pageSize }: ListParams) =>
  useQuery({
    queryKey: ['quotes', { search, status, partnerId, page, pageSize }],
    queryFn: async () => {
      const res = await api.get('/sales/quotes', {
        params: { search, status, partner_id: partnerId, page, page_size: pageSize },
      })
      return res.data as {
        items: Quote[]
        total: number
        page: number
        page_size: number
      }
    },
  })

export const getQuote = (id: string) =>
  useQuery({
    queryKey: ['quotes', id],
    queryFn: async () => {
      const res = await api.get(`/sales/quotes/${id}`)
      return res.data as QuoteDetail
    },
    enabled: !!id,
  })

export const createQuote = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (payload: QuoteCreate) =>
      api.post('/sales/quotes', payload).then((res) => res.data as QuoteDetail),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['quotes'] })
    },
  })
}

export const updateQuote = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ id, payload }: { id: string; payload: QuoteUpdate }) =>
      api.put(`/sales/quotes/${id}`, payload).then((res) => res.data as QuoteDetail),
    onSuccess: (_data, vars) => {
      queryClient.invalidateQueries({ queryKey: ['quotes'] })
      queryClient.invalidateQueries({ queryKey: ['quotes', vars.id] })
    },
  })
}

export const setQuoteStatus = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ id, status }: { id: string; status: string }) =>
      api.post(`/sales/quotes/${id}/status`, { status }).then((res) => res.data as QuoteDetail),
    onSuccess: (_data, vars) => {
      queryClient.invalidateQueries({ queryKey: ['quotes'] })
      queryClient.invalidateQueries({ queryKey: ['quotes', vars.id] })
    },
  })
}

export const convertQuoteToOrder = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      api.post(`/sales/quotes/${id}/to-order`).then((res) => res.data as any),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['quotes'] })
    },
  })
}

